### ISSUE TITLE
Upload a file

### ISSUE THREAD

**Issue by molexx** [jenkinsci/slack-plugin#181]
I'd like to be able to upload an artifact file from the workspace, I have it sort-of-working.

Unfortunately you can't upload files to slack using their 'webhooks', you need to use their 'web API', and you can't call that using Jenkins-CI's integration token, you need an oauth2 authtoken.

I'm not sure how this plugin is integrated into Slack's apps site - their documentation says that apps need their own server which is redirected to as part of the user setup/auth and can obtain the oauth2 authtoken. Is this happening for Jenkins-CI?

I note that 'view source' on the Jenkins-CI token page reveals an api_token in a javascript variable - which could be made visible on the page - but that seems to be the user's slack token which posts as them that they might not want to make available in their Jenkins config.

**Comment by molexx** [comment]
A workaround is to create a bot user as the bot gets its own authtoken which can call Slack's Web API.

It seems to be working OK, code is below, feedback appreciated.
https://github.com/molexx/slack-plugin

The create-a-bot process is described in README-UPLOAD.md.

I can submit a PR if there's interest and the extra authtoken field is considered acceptable, otherwise I'm happy to investigate alternative suggestions.

**Comment by dskrvk** [comment]
There's also a separate plugin for this, but it would be ideal to merge its functionality here: https://github.com/jenkinsci/slack-uploader-plugin

**Comment by timja** [comment]
Happy for this to be merged into the main plugin, would take a bit of cleanup work and shouldn't need the separate field as the slack-plugin has bot user support it would just only be supported under that mode

It also should have pipeline support which I don't think the fork has.

**Comment by IdanAdar** [comment]
So I didn't fully understand ho to accomplish this?

I have a bot user and I'd like it to attach a file to a thread. Specifically the Jenkins build log.

The file is, in theory, generated like so:
```
def logContent = Jenkins.getInstance().getItemByFullName(env.JOB_NAME).getBuildByNumber(Integer.parseInt(env.BUILD_NUMBER)).logFile.text
          writeFile file: "buildlog.txt", text: logContent
```

Now I'd like my Slack bot user to attach this file to a thread.
Replying to a thread was fixed in 2.26, so the first part of replying to a thread now works.

**Comment by timja** [comment]
I assume you need to use this API?
https://api.slack.com/methods/files.upload
but I haven't looked too closely

**Comment by molexx** [comment]
Unfortunately I don't have time to merge and update as suggested my fork, anyone is welcome to pick it up:

https://github.com/molexx/slack-plugin/commit/7c89059417a248a4ce4afa715d7ad0ab2e15e761

**Comment by kurekbharath** [comment]
> So I didn't fully understand ho to accomplish this?
> 
> I have a bot user and I'd like it to attach a file to a thread. Specifically the Jenkins build log.
> 
> The file is, in theory, generated like so:
> 
> ```
> def logContent = Jenkins.getInstance().getItemByFullName(env.JOB_NAME).getBuildByNumber(Integer.parseInt(env.BUILD_NUMBER)).logFile.text
> writeFile file: "buildlog.txt", text: logContent
> ```
> 
> Now I'd like my Slack bot user to attach this file to a thread. Replying to a thread was fixed in 2.26, so that part works, just need to attach the file to the reply...

Is the above snippet not encoding only a FEW strings from the consoleouptut when it is written to a text file? I am trying same approach for me it is encoding few string please confirm if it is happening to you also?
